File: iosApp/iosApp/Pages/Search/Results/SearchResultsView.swift
//
//  SearchResultsView.swift
//  iosApp
//
//  Created by Brandon Rodriguez on 10/9/24.
//  Copyright © 2024 MBTA. All rights reserved.
//

import Foundation
import Shared
import SwiftUI

struct SearchResultsView: View {
    private var state: SearchViewModel.State
    private var handleStopTap: (String) -> Void

    @EnvironmentObject var settingsCache: SettingsCache

    var includeRoutes: Bool { settingsCache.get(.searchRouteResults) }

    init(
        state: SearchViewModel.State,
        handleStopTap: @escaping (String) -> Void
    ) {
        self.state = state
        self.handleStopTap = handleStopTap
    }

    var body: some View {
        ScrollView {
            Group {
                switch onEnum(of: state) {
                case .loading:
                    LoadingResultsView()
                case let .recentStops(state):
                    VStack {
                        Text("Recently Viewed")
                            .font(Typography.subheadlineSemibold)
                            .frame(maxWidth: .infinity, alignment: .leading)
                        StopResultsView(stops: state.stops, handleStopTap: handleStopTap)
                    }
                case let .results(state):
                    if state.isEmpty(includeRoutes: includeRoutes) {
                        EmptyStateView(
                            headline: NSLocalizedString(
                                "No results found 🤔",
                                comment: "Displayed when search has no results"
                            ),
                            subheadline: NSLocalizedString(
                                "Try a different spelling or name.",
                                comment: "Displayed when search has no results"
                            ),
                        )
                        .padding(.top, 16)
                    } else {
                        VStack(spacing: 8) {
                            StopResultsView(stops: state.stops, handleStopTap: handleStopTap)
                            if includeRoutes, !state.routes.isEmpty {
                                RouteResultsView(routes: state.routes)
                                    .padding(.top, 8)
                            }
                        }
                    }
                case .error:
                    EmptyStateView(
                        headline: NSLocalizedString(
                            "Results failed to load ☹️",
                            comment: "Displayed when search encounters an error"
                        ),
                        subheadline: NSLocalizedString(
                            "Try your search again.",
                            comment: "Displayed when search encounters an error"
                        ),
                    )
                    .padding(.top, 16)
                }
            }
            .onChange(of: state) { state in
                if case let .results(state) = onEnum(of: state), !state.isEmpty(includeRoutes: includeRoutes) {
                    let announcementString = String(format: NSLocalizedString(
                        "%ld results found",
                        comment: "Screen reader text that is announced when search results are returned"
                    ), state.stops.count)

                    if #available(iOS 17, *) {
                        var resultsFoundAnnouncement = AttributedString(announcementString)
                        resultsFoundAnnouncement.accessibilitySpeechAnnouncementPriority = .high
                        AccessibilityNotification.Announcement(resultsFoundAnnouncement).post()
                    } else {
                        UIAccessibility.post(
                            notification: .layoutChanged,
                            argument: announcementString
                        )
                    }
                }
            }
            .animation(.easeInOut(duration: 0.25), value: state)
            .frame(maxWidth: .infinity, alignment: .topLeading)
            .padding(.vertical, 16)
            .padding(.horizontal, 16)
        }
        .background(Color.fill1)
    }
}

File: androidApp/src/main/java/com/mbta/tid/mbta_app/android/search/results/SearchResultsView.kt
package com.mbta.tid.mbta_app.android.search.results

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.navigationBarsPadding
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.res.colorResource
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.unit.dp
import com.mbta.tid.mbta_app.android.R
import com.mbta.tid.mbta_app.android.util.SettingsCache
import com.mbta.tid.mbta_app.android.util.Typography
import com.mbta.tid.mbta_app.model.LineOrRoute
import com.mbta.tid.mbta_app.repositories.Settings
import com.mbta.tid.mbta_app.viewModel.SearchViewModel

@Composable
fun SearchResultsView(
    state: SearchViewModel.State,
    handleStopTap: (String) -> Unit,
    handleRouteTap: (LineOrRoute.Id) -> Unit,
) {
    val includeRoutes = SettingsCache.get(Settings.SearchRouteResults)
    Column(
        modifier =
            Modifier.verticalScroll(rememberScrollState())
                .fillMaxWidth()
                .background(colorResource(R.color.fill1))
                .padding(16.dp)
                .navigationBarsPadding()
    ) {
        when (state) {
            SearchViewModel.State.Loading -> LoadingResultsView()
            is SearchViewModel.State.RecentStops -> {
                Column(verticalArrangement = Arrangement.spacedBy(8.dp)) {
                    Text(
                        modifier = Modifier.padding(bottom = 10.dp),
                        text = stringResource(R.string.recently_viewed),
                        style = Typography.subheadlineSemibold,
                    )
                    StopResultsView(state.stops, handleStopTap)
                }
            }
            is SearchViewModel.State.Results -> {
                if (state.isEmpty(includeRoutes)) {
                    EmptyStateView(
                        headline = stringResource(R.string.no_results_found),
                        subheadline = stringResource(R.string.try_a_different_spelling_or_name),
                        modifier = Modifier.padding(top = 16.dp).fillMaxWidth(),
                    )
                } else {
                    Column(verticalArrangement = Arrangement.spacedBy(8.dp)) {
                        StopResultsView(state.stops, handleStopTap)
                        if (includeRoutes && !state.routes.isEmpty()) {
                            RouteResultsView(state.routes, handleRouteTap)
                        }
                    }
                }
            }
            SearchViewModel.State.Error ->
                EmptyStateView(
                    headline = stringResource(R.string.results_failed_to_load),
                    subheadline = stringResource(R.string.try_your_search_again),
                    modifier = Modifier.padding(top = 16.dp).fillMaxWidth(),
                )
        }
    }
}
